internal/nozzle/nozzle.go

```diff
@@ -112,63 +113,74 @@ func (d *Nozzle) Start() error {
+	err = d.startFirehoseConsumer(authToken)
+	if err != nil {
+		return err
+	}
 
-		// Start multiple workers to parallelize firehose events (event.envelope) transformation into processedMetrics
-		// and then grouped into metricsMap
-		d.startWorkers()
-		// Start infinite loop to periodically submit metrics (metricsMap) to Datadog
-		err = d.postToDatadog()
+	// Start multiple workers to parallelize firehose events (event.envelope) transformation into processedMetrics
+	// and then grouped into metricsMap
+	d.startWorkers()
 
-		if !shouldReconnect(err) {
-			break
-		}
+	// Execute infinite loop.
+	// This method is blocking until we get error or a stop signal
+	err = d.Run()
 
-		// memorize the retry timestamp
-		retryHistory = append(retryHistory[1:], time.Now())
-		if retryHistory[2].Sub(retryHistory[0]) < 5*time.Second {
-			// We retried three times quickly, there might be a larger issue
-			// Let's break out of the loop and return the error
-			d.log.Info("Too many retries, shutting down...")
-			break
-		}
-		d.stopWorkers()
-		// Sleep a little to not reconnect to quickly
-		time.Sleep(500 * time.Millisecond)
-		d.log.Info("Websocket connection lost, reestablishing connection...")
-	}
-
-	// Whenever a stop signal is received the infinite loop within postToDatadog is stopped and then we stop all workers
+	// Whenever a stop signal is received the Run methode above will return. The code below will then be executed
 	d.log.Info("DataDog Firehose Nozzle shutting down...")
+	// Close Firehose Consumer
+	d.log.Infof("Closing connection with traffic controller due to %v", err)
+	d.consumer.Close()
+	// stop processor
 	d.stopWorkers()
+	// Submit metrics left in cache if any
+	d.PostMetrics()
 
 	return err
 }
 
-// Define the error cases where we should reattempt a connection
-func shouldReconnect(err error) bool {
-	if err == nil {
-		return false
+func (n *Nozzle) startFirehoseConsumer(authToken string) error {
+	var err error
+	// Initialize the firehose consumer (with retry enable)
+	n.consumer, err = n.newFirehoseConsumer(authToken)
+	if err != nil {
+		return err
 	}
-	switch err.(type) {
-	case noaaerrors.RetryError:
-		return true
-	case noaaerrors.NonRetryError:
-		return false
-	default:
-		return false
+	// Run the Firehose consumer
+	// It consumes messages from the Firehose and push them to n.messages
+	n.messages, n.errors = n.consumer.FilteredFirehose(n.config.FirehoseSubscriptionID, authToken, consumer.Metrics)
+	return nil
+}
+
+func (d *Nozzle) Run() error {
+	// Start infinite loop to periodically:
+	// - submit metrics to Datadog
+	// - handle error
+	//   - log error
+	//   - break out of the loop if error is not a retry error
+	// - stop nozzle
+	//   - break out of the loop
+	ticker := time.NewTicker(time.Duration(d.config.FlushDurationSeconds) * time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			// Submit metrics to Datadog
+			d.PostMetrics()
+		case error := <-d.errors:
```

Comment:
let's name the var `err`, `error` is a type

Reply:
Actually i fixed that already on my local :)